### PR TITLE
Unify attribute modification API

### PR DESCRIPTION
**Pull request by jjohannes**

### CODE REVIEW

File: subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/configurations/DefaultConfiguration.java [hunk 1]
```diff
@@ -459,6 +460,8 @@ public void execute(BuildOperationContext buildOperationContext) {
                 dependencyResolutionListeners.getSource().afterResolve(incoming);
                 // Discard listeners
                 dependencyResolutionListeners.removeAll();
+
+                lockAttributes();
```
Comment (oehme) [inline subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/configurations/DefaultConfiguration.java]: Shouldn't we lock them before resolution?

Comment (jjohannes) [inline subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/configurations/DefaultConfiguration.java]: Makes sense. I moved it up to the beginning of the _resolve_ code block.

Note that lockAttributes is already called earlier in a resolution process here:
https://github.com/gradle/gradle/blob/7eb9d8b17d1e8e0e16d945dbc4300137dca194c6/subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/moduleconverter/DefaultConfigurationComponentMetaDataBuilder.java#L47-L47

I added this to make sure a lock is performed when `resolve()` is called on a configuration from another place. I was triggered to do this by this test:
https://github.com/gradle/gradle/blob/f61eb3671920b388f505b7c5ebacc4b95f99457b/subprojects/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/UnsupportedConfigurationMutationTest.groovy#L455-L455

File: subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/configurations/DefaultConfiguration.java [hunk 2]
```diff
@@ -991,9 +970,16 @@ public ArtifactView artifactView() {
...
-            public ArtifactView attributes(Map<?, ?> attributeMap) {
+            public AttributeContainer getAttributes() {
+                return getViewAttributes();
+            }
+
+            @Override
+            public ArtifactView attributes(Action<? super AttributeContainer> action) {
                 assertUnset("attributes", viewAttributes);
-                this.viewAttributes = attributesFactory.fromPolymorphicMap(attributeMap);
+                DefaultMutableAttributeContainer mutableAttributes = new DefaultMutableAttributeContainer(attributesFactory);
```
Comment (oehme) [inline subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/configurations/DefaultConfiguration.java]: This is inconsistent with Configurations. I should be able to call this as many times as I wish and it should be locked once I call `getFiles` or `getArtifacts`

File: subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/configurations/DefaultConfiguration.java [hunk 3]
```diff
@@ -591,7 +592,50 @@ public ImmutableAttributesFactory getAttributesCache() {
 
     @Override
     public void lockAttributes() {
-        configurationAttributes = configurationAttributes.asImmutable();
+        final AttributeContainerInternal delegatee = configurationAttributes.asImmutable();
+        configurationAttributes = new AttributeContainerInternal() {
+
+            @Override
+            public ImmutableAttributes asImmutable() {
+                return delegatee.asImmutable();
```
Comment (melix) [inline subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/configurations/DefaultConfiguration.java]: You could return `this` instead

File: subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/configurations/DefaultConfiguration.java (hunk 2 above)
Comment (jjohannes) [inline subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/configurations/DefaultConfiguration.java]: I was thinking that we may address this another time. But you are right and it bothers me too. I will do something on this.

Comment (jjohannes) [inline subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/configurations/DefaultConfiguration.java]: To be consistent, `getAttributes()` should also always return the mutable container before `getFiles` or `getArtifacts` is called. Atm it returns `EMPTY` for an empty container.

Comment (jjohannes) [inline subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/configurations/DefaultConfiguration.java]: Done.

File: subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/configurations/DefaultConfiguration.java [hunk 4]
```diff
@@ -591,7 +592,50 @@ public ImmutableAttributesFactory getAttributesCache() {
 
     @Override
     public void lockAttributes() {
-        configurationAttributes = configurationAttributes.asImmutable();
+        final AttributeContainerInternal delegatee = configurationAttributes.asImmutable();
+        configurationAttributes = new AttributeContainerInternal() {
```
Comment (oehme) [inline subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/configurations/DefaultConfiguration.java]: Might be worth extracting this into a private inner class to keep the method short.

File: subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/configurations/DefaultConfiguration.java [hunk 5]
```diff
@@ -989,46 +1012,57 @@ public ArtifactView artifactView() {
...
+                assertComponentFilterUnset();
                 this.viewFilter = componentFilter;
                 return this;
             }
 
-            private void assertUnset(String method, Object value) {
-                if (value != null) {
-                    throw new IllegalStateException("Cannot call " + method + " multiple times for view");
+            private void assertComponentFilterUnset() {
+                if (viewFilter != null) {
```
Comment (oehme) [inline subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/configurations/DefaultConfiguration.java]: Should be componentFilter

Comment (jjohannes) [inline subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/configurations/DefaultConfiguration.java]: done

File: subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/configurations/DefaultConfiguration.java (hunk 4 above)
Comment (jjohannes) [inline subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/configurations/DefaultConfiguration.java]: done

**Comment by jjohannes** [comment]
Successful Branch Build: https://builds.gradle.org/viewLog.html?buildId=2167793

**Pull request merged**